site/docs/tutorial/sql_to_substrait.md

```diff
@@ -0,0 +1,1305 @@
+adds new fields, which are visible to it's downstream relations. The diagram
+below shoes the mapping of field indices within each relation and how each of
+the field references show up in each relations properties.
+
+![](field_indices_layout.svg)
+
+## Column selection and emit
+
+Looking at the relations so far, you might be wondering: when do we select the
+output columns?
+
+You might be tempted to add a [Project relation](/relations/logical_relations/#project-operation) 
+at the end to select a subset of columns. However, the project relation only
+adds columns; it is not responsible for subsetting or reordering columns.
+
+<!-- TODO: Is this true? -->
+Instead, any relation can reorder or subset columns through the `emit` property.
+By default, it is set to `direct`, which outputs all columns "as is". But it can
+also be specified as a sequence of field indices.
+
+For simplicity, we will add this to the final aggregate relation. We could add it
+to all relations, only selecting the fields we strictly need in later relations.
+Indeed, a good optimizer would probably do that to our plan. But to keep things
+simple, we'll limit the columns at the end.
+
+For our final output, we'll modify to be:
+
+```json
+{
+  "aggregate": {
+    "input": { ... },
+    "groupings": [ ... ],
+    "measures": [ ... ],
+    "emit_type": {
+      "emit": [9, 0, 10]
+    }
+}
+```
+
+Notice, we are using index 10 to reference the column calculated during the
+aggregation.
+
+## Plans
+
+Now that we've constructed our relations, we can put it all into a plan.
+Substrait plans are the only messages that can be sent and received on their
+own. Recall that earlier, we had function references to those YAML files, but
+so far there's been no place to tell a consumer what those function reference
+IDs mean or which extensions we are using. That information, goes at the plan
+level.
+
+The overall layout for a plan is
+
+```json
+{
+  "extensionUris": [ ... ],
+  "extensions": [ ... ],
+  "relations": [
+    {
+      "root": {
+        "names": [
+          "product_name",
+          "product_id",
+          "sales"
+        ],
+        "input": { ... }
+      }
+    }
+  ]
+}
+```
+
+The `relations` field is a list of Root relations, which provides the final
+column names for our query. The input to this relation is our aggregate relation
+(which contains all the other relations as children).
+
+For extensions, we need to provide `extensionUris` with the locations of the 
+YAML files we used and `extensions` with the list of functions we used and which
+extension they come from.
+
+In our query, we used:
+
+ * `index_in` (0), from `functions_string.yaml`,
+ * `is_null` (1), from `functions_comparison.yaml`,
+ * `equal` (2), from `functions_comparison.yaml`,
+ * `sum` (3), from `functions_arithmetic_decimal.yaml`,
+ * `multiply` (4), from `functions_arithmetic_decimal.yaml`.
+
+So first we can create the three extension uris:
+
+<!-- TODO: what is the URI format? -->
```

Comment:
All the plans I have seen from ibis-substrait omits the URI, but that seems... wrong? Have we established the pattern for these URIs?

Reply:
> Have we established the pattern for these URIs?

Not quite, but the time is ripe to fix this.  All produers I've seen (Isthmus & DuckDb) operate the same as Ibis.  Originally the contention was between:

 * Brittle URIs that are tied to a specific version of an extension function but the content is fixed (e.g. https://github.com/substrait-io/substrait/blob/v0.18.0/extensions/functions_arithmetic.yaml)
 * Stable URIs that are not tied to a specific version and the content may vary (e.g. https://github.com/substrait-io/substrait/blob/main/extensions/functions_arithmetic.yaml)

Based on the conversation in https://github.com/substrait-io/substrait/issues/274 I believe we were leaning towards the latter (stable URI, potentially changing content).

Note that, in Acero, we treat the URI `/` (or the empty string) as a special case where we fallback to name-only lookup and look for a function with the given name, regardless of what URI it maps to (and if there are multiple I think we pick the first we find).  This allows us to consume plans from producers today that omit the URI.